lib/efi_services/efi_services.c

```diff
@@ -35,3 +36,49 @@ boolean_t init_efi_services(uint64_t sys_table)
 
 	return TRUE;
 }
+
+uint64_t efi_get_tom(void)
+{
+        uintn_t mmap_size, map_key, desc_size;
+        uint32_t desc_ver;
+	uint64_t end;
+        uint64_t tom = 0;
+        efi_memory_descriptor_t temp_mmap[1];
+        efi_memory_descriptor_t *mmap_desc;
+        efi_memory_descriptor_t *mmap_desc_ptr;
+        efi_status_t status;
+        uintn_t i;
+
+        /* get mmap_size */
+        mmap_size = sizeof(temp_mmap);
+        status = g_bs->get_memory_map(&mmap_size, &temp_mmap[0], &map_key, &desc_size, &desc_ver);
+
+        /* allocate space for mmap */
+        mmap_size += PAGE_4K_SIZE;
+        status = g_bs->allocate_pool(EfiLoaderData, mmap_size, (void **)&mmap_desc);
+        if (status != EFI_SUCCESS) {
+                return 0;
+        }
+
+        /* get mmap */
+        status = g_bs->get_memory_map(&mmap_size, mmap_desc, &map_key, &desc_size, &desc_ver);
+        if (status != EFI_SUCCESS) {
+                return 0;
+        }
+
+	/* caculate top of memory */
+        mmap_desc_ptr = mmap_desc;
+        for (i = 0; i < mmap_size/desc_size; i++) {
+		end = mmap_desc_ptr->physical_start + mmap_desc_ptr->number_of_pages * PAGE_4K_SIZE;
+                if (tom < end) {
+                        tom = end;
+                }
+
+                mmap_desc_ptr = (efi_memory_descriptor_t *)(((uintn_t)mmap_desc_ptr) + desc_size);
+        }
+
+	/* Prevent BIOS interrupts */
+	asm volatile("cli;");
```

Comment:
where's the "sti"?

Reply:
sti is called in efi api. do we need to "sti" explicitly in this function early?